### DISCUSSION TITLE
Metrics Server not available on a fresh install

### DISCUSSION THREAD

**Discussion by Metamogul** [k3s-io/k3s#7132] [Q&A] [answered]
**Environmental Info:**
K3s Version: 

    k3s version v1.25.7+k3s1 (f7c20e23)
    go version go1.19.6

Node(s) CPU architecture, OS, and Version: 

    Linux virtualpi3 5.10.0-21-arm64 #1 SMP Debian 5.10.162-1 (2023-01-21) aarch64 GNU/Linux

Cluster Configuration:

1 Single Server

**Describe the bug:**
When installing a k3s server via `curl -sfL https://get.k3s.io | sh -s - --write-kubeconfig-mode 644` on a fresh debian-based system, the metrics server seems to be unavailabe. This is what `kubectl get apiservices` returns:

    NAME                                   SERVICE                      AVAILABLE                   AGE
    v1.admissionregistration.k8s.io        Local                        True                        12s
    v1.                                    Local                        True                        12s
    v1.authentication.k8s.io               Local                        True                        12s
    v1.apiextensions.k8s.io                Local                        True                        12s
    v1.apps                                Local                        True                        12s
    v1.autoscaling                         Local                        True                        12s
    v1.batch                               Local                        True                        12s
    v2.autoscaling                         Local                        True                        12s
    v1.authorization.k8s.io                Local                        True                        12s
    v1.certificates.k8s.io                 Local                        True                        12s
    v1.discovery.k8s.io                    Local                        True                        12s
    v1.events.k8s.io                       Local                        True                        12s
    v2beta2.autoscaling                    Local                        True                        12s
    v1.coordination.k8s.io                 Local                        True                        12s
    v1beta2.flowcontrol.apiserver.k8s.io   Local                        True                        12s
    v1.node.k8s.io                         Local                        True                        12s
    v1.networking.k8s.io                   Local                        True                        12s
    v1beta1.flowcontrol.apiserver.k8s.io   Local                        True                        12s
    v1.policy                              Local                        True                        12s
    v1.rbac.authorization.k8s.io           Local                        True                        12s
    v1beta1.storage.k8s.io                 Local                        True                        12s
    v1.scheduling.k8s.io                   Local                        True                        12s
    v1.storage.k8s.io                      Local                        True                        12s
    v1.k3s.cattle.io                       Local                        True                        10s
    v1.helm.cattle.io                      Local                        True                        10s
    v1beta1.metrics.k8s.io                 kube-system/metrics-server   False (EndpointsNotFound)   7s

Subsequent commands that rely on the metrics server like `kubectl top nodes` fail, the given example with the output

    Error from server (ServiceUnavailable): the server is currently unable to handle the request (get nodes.metrics.k8s.io)

**Steps To Reproduce:**

Install k3s via the command provided above on a fresh Dietpi install in a VM. 

**Expected behavior:**

After the given install command I would expect k3s to install the metrics server that was not disabled.

**Actual behavior:**

The metrics server seems to be not reachable. This leads to the failure of other components relying on it. Since this is my starting point to teach myself the k8s framework, I don't have the slightest clue what has gone wrong here and how I could fix it.

**Additional context / logs:**

Output from install command provided above:

    [INFO]  Finding release for channel stable
    [INFO]  Using v1.25.7+k3s1 as release
    [INFO]  Downloading hash https://github.com/k3s-io/k3s/releases/download/v1.25.7+k3s1/sha256sum-arm64.txt
    [INFO]  Downloading binary https://github.com/k3s-io/k3s/releases/download/v1.25.7+k3s1/k3s-arm64
    [INFO]  Verifying binary download
    [INFO]  Installing k3s to /usr/local/bin/k3s
    [INFO]  Skipping installation of SELinux RPM
    [INFO]  Creating /usr/local/bin/kubectl symlink to k3s
    [INFO]  Creating /usr/local/bin/crictl symlink to k3s
    [INFO]  Creating /usr/local/bin/ctr symlink to k3s
    [INFO]  Creating killall script /usr/local/bin/k3s-killall.sh
    [INFO]  Creating uninstall script /usr/local/bin/k3s-uninstall.sh
    [INFO]  env: Creating environment file /etc/systemd/system/k3s.service.env
    [INFO]  systemd: Creating service file /etc/systemd/system/k3s.service
    [INFO]  systemd: Enabling k3s unit
    Created symlink /etc/systemd/system/multi-user.target.wants/k3s.service → /etc/systemd/system/k3s.service.
    [INFO]  systemd: Starting k3s

**Comment by brandond** [comment]
Have you checked the actual Metrics Server pod logs to see why it's not available? You've checked everything except the actual application hosting the service.

**Comment by Metamogul** [comment]
The output of `kubectl logs metrics-server-5f9f776df5-ncr9b -n kube-system` returns

    Error from server (BadRequest): container "metrics-server" in pod "metrics-server-5f9f776df5-ncr9b" is waiting to start: CreateContainerError

I looks like the pod couldn't be created in the first place.

**Reply by brandond** [reply]
Did you remember to specify the namespace?

**Reply by Metamogul** [reply]
Actually I didn't, thank you very much for the hint! I've updated my post with the correct command & output.

**Comment by Metamogul** [comment]
So, here comes an update: Running the exact same installation procedure on actual hardware (`Linux raspberrypi3 6.1.19-v8+ #1637 SMP PREEMPT Tue Mar 14 11:11:47 GMT 2023 aarch64 GNU/Linux`) leads to the expected behaviour and doesn't yield any issues. It seems that the bug is related to the specific environment. The VM image in question has been created by running the Dietpi installation procedure against a recent Debian netinstall image https://cdimage.debian.org/debian-cd/current/arm64/iso-cd/debian-11.6.0-arm64-netinst.iso. It seemed to me that this was the closest to the actual hardware I want to provision; yet apparently there are some key differences for the given use case I fail to understand.

The output of `kubectl get pods -n kube-system` on the reported VM is as follows:

    NAME                                      READY   STATUS                 RESTARTS   AGE
    local-path-provisioner-79f67d76f8-8hffn   0/1     CreateContainerError   0          145m
    helm-install-traefik-crd-5h9d2            0/1     CreateContainerError   0          145m
    metrics-server-5f9f776df5-ncr9b           0/1     CreateContainerError   0          145m
    coredns-597584b69b-rmg2k                  0/1     CreateContainerError   0          145m
    helm-install-traefik-7cbzk                0/1     CreateContainerError   0          145m

Whereas on the physical machine mentioned above where everything runs as expected:

    NAME                              READY   STATUS      RESTARTS   AGE
    coredns-597584b69b-zh5b9          1/1     Running     0          107s
    metrics-server-5f9f776df5-jpxj5   1/1     Running     0          107s
    helm-install-traefik-crd-2nfvv    0/1     Completed   0          108s
    helm-install-traefik-g7kkv        0/1     Completed   1          108s
    traefik-66c46d954f-vlbtd          1/1     Running     0          73s

The output of the installation is completely identical to the installation on the VM.

Following a debug guide for pods failing with the `CreateContainerError` message I've also looked into `kubectl describe pod metrics-server-5f9f776df5-ncr9b -n kube-system` and it returned:

    Name:                 metrics-server-5f9f776df5-ncr9b
    Namespace:            kube-system
    Priority:             2000001000
    Priority Class Name:  system-node-critical
    Service Account:      metrics-server
    Node:                 virtualpi3/192.168.205.22
    Start Time:           Tue, 21 Mar 2023 19:19:15 +0100
    Labels:               k8s-app=metrics-server
                          pod-template-hash=5f9f776df5
    Annotations:          <none>
    Status:               Pending
    IP:                   10.42.0.9
    IPs:
      IP:           10.42.0.9
    Controlled By:  ReplicaSet/metrics-server-5f9f776df5
    Containers:
      metrics-server:
        Container ID:
        Image:         rancher/mirrored-metrics-server:v0.6.2
        Image ID:
        Port:          10250/TCP
        Host Port:     0/TCP
        Args:
          --cert-dir=/tmp
          --secure-port=10250
          --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
          --kubelet-use-node-status-port
          --metric-resolution=15s
        State:          Waiting
          Reason:       CreateContainerError
        Ready:          False
        Restart Count:  0
        Requests:
          cpu:        100m
          memory:     70Mi
        Liveness:     http-get https://:https/livez delay=60s timeout=1s period=10s #success=1 #failure=3
        Readiness:    http-get https://:https/readyz delay=0s timeout=1s period=2s #success=1 #failure=3
        Environment:  <none>
        Mounts:
          /tmp from tmp-dir (rw)
          /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-drzv2 (ro)
    Conditions:
      Type              Status
      Initialized       True
      Ready             False
      ContainersReady   False
      PodScheduled      True
    Volumes:
      tmp-dir:
        Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
        Medium:
        SizeLimit:  <unset>
      kube-api-access-drzv2:
        Type:                    Projected (a volume that contains injected data from multiple sources)
        TokenExpirationSeconds:  3607
        ConfigMapName:           kube-root-ca.crt
        ConfigMapOptional:       <nil>
        DownwardAPI:             true
    QoS Class:                   Burstable
    Node-Selectors:              <none>
    Tolerations:                 CriticalAddonsOnly op=Exists
                                 node-role.kubernetes.io/control-plane:NoSchedule op=Exists
                                 node-role.kubernetes.io/master:NoSchedule op=Exists
                                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
    Events:
      Type    Reason  Age                    From     Message
      ----    ------  ----                   ----     -------
      Normal  Pulled  78s (x415 over 3h18m)  kubelet  Container image "rancher/mirrored-metrics-server:v0.6.2" already present on machine

Btw. thank you very much for taking the time to investigate. I really appreciate that.

**Reply by brandond** [reply]
I would check the containerd and k3s error logs to see if there's anything interesting in there.

You have seen the notes about raspberry pi in the docs, correct?
https://docs.k3s.io/advanced#raspberry-pi

**Reply by Metamogul** [reply]
Thanks for pointing me there. I've considered that and included a task in my playbook for bootstrapping the physical machines, as I've made sure cgroups are active on the VM.

**Comment by Metamogul** [comment, marked as answer]
The output of `sudo journalctl -u containerd` got me nowhere, but the k3s error log is full of error messages, the most recent ones:

      ...
      Mar 22 00:08:22 virtualpi3 k3s[507]: E0322 00:08:22.671479     507 remote_runtime.go:442] "CreateContainer in sandbox from runtime service failed" err="rpc error: code = Unknown desc = failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)" podSandboxID="de358637ae6d82b9be950133e870245812369330bfa34e3a7ec3336d97a43e1e"
      Mar 22 00:08:22 virtualpi3 k3s[507]: E0322 00:08:22.671556     507 kuberuntime_manager.go:862] container &Container{Name:helm,Image:rancher/klipper-helm:v0.7.4-build20221121,Command:[],Args:[install --set-string global.systemDefaultRegistry=],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{EnvVar{Name:NAME,Value:traefik,ValueFrom:nil,},EnvVar{Name:VERSION,Value:,ValueFrom:nil,},EnvVar{Name:REPO,Value:,ValueFrom:nil,},EnvVar{Name:HELM_DRIVER,Value:secret,ValueFrom:nil,},EnvVar{Name:CHART_NAMESPACE,Value:kube-system,ValueFrom:nil,},EnvVar{Name:CHART,Value:https://%{KUBERNETES_API}%/static/charts/traefik-20.3.1+up20.3.0.tgz,ValueFrom:nil,},EnvVar{Name:HELM_VERSION,Value:,ValueFrom:nil,},EnvVar{Name:TARGET_NAMESPACE,Value:kube-system,ValueFrom:nil,},EnvVar{Name:NO_PROXY,Value:.svc,.cluster.local,10.42.0.0/16,10.43.0.0/16,ValueFrom:nil,},EnvVar{Name:FAILURE_POLICY,Value:reinstall,ValueFrom:nil,},},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[]VolumeMount{VolumeMount{Name:values,ReadOnly:false,MountPath:/config,SubPath:,MountPropagation:nil,SubPathExpr:,},VolumeMount{Name:content,ReadOnly:false,MountPath:/chart,SubPath:,MountPropagation:nil,SubPathExpr:,},VolumeMount{Name:kube-api-access-ldbfj,ReadOnly:true,MountPath:/var/run/secrets/kubernetes.io/serviceaccount,SubPath:,MountPropagation:nil,SubPathExpr:,},},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:/dev/termination-log,ImagePullPolicy:IfNotPresent,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:File,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,} start failed in pod helm-install-traefik-7cbzk_kube-system(e5044688-f66e-4a0f-ba97-0e2dfddd58c4): CreateContainerError: failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)
      Mar 22 00:08:22 virtualpi3 k3s[507]: E0322 00:08:22.671576     507 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"helm\" with CreateContainerError: \"failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)\"" pod="kube-system/helm-install-traefik-7cbzk" podUID=e5044688-f66e-4a0f-ba97-0e2dfddd58c4
      Mar 22 00:08:23 virtualpi3 k3s[507]: E0322 00:08:23.669842     507 remote_runtime.go:442] "CreateContainer in sandbox from runtime service failed" err="rpc error: code = Unknown desc = failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)" podSandboxID="aa53f5cad2f828832335bb50c2be155c1d83450c8b799dda8a5f885db9411890"
      Mar 22 00:08:23 virtualpi3 k3s[507]: E0322 00:08:23.669994     507 kuberuntime_manager.go:862] container &Container{Name:helm,Image:rancher/klipper-helm:v0.7.4-build20221121,Command:[],Args:[install],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{EnvVar{Name:NAME,Value:traefik-crd,ValueFrom:nil,},EnvVar{Name:VERSION,Value:,ValueFrom:nil,},EnvVar{Name:REPO,Value:,ValueFrom:nil,},EnvVar{Name:HELM_DRIVER,Value:secret,ValueFrom:nil,},EnvVar{Name:CHART_NAMESPACE,Value:kube-system,ValueFrom:nil,},EnvVar{Name:CHART,Value:https://%{KUBERNETES_API}%/static/charts/traefik-crd-20.3.1+up20.3.0.tgz,ValueFrom:nil,},EnvVar{Name:HELM_VERSION,Value:,ValueFrom:nil,},EnvVar{Name:TARGET_NAMESPACE,Value:kube-system,ValueFrom:nil,},EnvVar{Name:NO_PROXY,Value:.svc,.cluster.local,10.42.0.0/16,10.43.0.0/16,ValueFrom:nil,},EnvVar{Name:FAILURE_POLICY,Value:reinstall,ValueFrom:nil,},},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[]VolumeMount{VolumeMount{Name:values,ReadOnly:false,MountPath:/config,SubPath:,MountPropagation:nil,SubPathExpr:,},VolumeMount{Name:content,ReadOnly:false,MountPath:/chart,SubPath:,MountPropagation:nil,SubPathExpr:,},VolumeMount{Name:kube-api-access-bdn2j,ReadOnly:true,MountPath:/var/run/secrets/kubernetes.io/serviceaccount,SubPath:,MountPropagation:nil,SubPathExpr:,},},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:/dev/termination-log,ImagePullPolicy:IfNotPresent,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:File,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,} start failed in pod helm-install-traefik-crd-5h9d2_kube-system(75d94752-526a-4bcc-835e-868109e5213b): CreateContainerError: failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)
      Mar 22 00:08:23 virtualpi3 k3s[507]: E0322 00:08:23.670242     507 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"helm\" with CreateContainerError: \"failed to create containerd container: get apparmor_parser version: apparmor_parser resolves to executable in current directory (./apparmor_parser)\"" pod="kube-system/helm-install-traefik-crd-5h9d2" podUID=75d94752-526a-4bcc-835e-868109e5213b

~~I'll try to make sence of that.~~ I could resolve the issue with that info and figured that I might need to install apparmor. A simple `sudo apt install apparmor` did it for me, not even a reboot was required. Thanks again for the help!
